### DISCUSSION TITLE
clipboard_file: cannot handle files more than 2GiB unintendedly and HUGE_FILE_SUPPORT

### DISCUSSION THREAD

**Discussion by metalefty** [neutrinolabs/xrdp#1987] [General]
I found that clipboard file transfer might not work properly if the filesize > 4GiB.

Also, there are some codes using `int` that might have values of more than 2^32 like #1087.

I started adding some tests for this.

https://github.com/neutrinolabs/xrdp/blob/b481351ad7440e0e05f11f2df5214e46171d4d0e/sesman/chansrv/clipboard_file.c#L53-L60
https://github.com/neutrinolabs/xrdp/blob/b481351ad7440e0e05f11f2df5214e46171d4d0e/sesman/chansrv/clipboard_file.c#L191
https://github.com/neutrinolabs/xrdp/blob/b481351ad7440e0e05f11f2df5214e46171d4d0e/common/os_calls.c#L2493-L2494

**Comment by matt335672** [comment]
Good point.

`int` is going to be signed, so potentially 2Gb rather than 4Gb isn't it?

What do you think about the idea of using `off_t` from `<sys/types.h>` for this? We're already using it for the FUSE stuff, as it's in the FUSE data structures.

We should also consider `g_file_seek()` in this. 

I think most uses could be found quite easily

1. Rename `g_file_get_size()` and `g_file_seek()` as `zzzz_g_file_get_size()` and `zzzz_g_file_seek()`
2. Fix these functions to work correctly with `off_t` (or whatever)
3. Recompile and fix references to `zzzz_`, changing types along the way. Where calls are assigned to variables or struct members, give these variables a prefix of `zzzz_` also.
4. Use github actions locally to check everything is OK.
5. When all is working, remove all the `zzzz_` with something like (I HAVEN'T TESTED THIS):-
```
filelist=$(find . -type f -exec grep zzzz_ {} +)
echo $filelist ; Check it looks sane
for file in $filelist; do sed -i -e 's/zzzz_//g'; ;done
```

Does that make sense? Will it work OK?

**Reply by metalefty** [reply]
Indeed, it is actually 2GiB.

Also, it sounds good to me to use `off_t` instead. 

The steps you show makes sense to me, I'll give it a try.

**Comment by metalefty** [comment]
I found that the file size transferred over the clipboard is limited to 4 GiB in protocol design. 

https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/7718c8c9-798d-4788-bb75-64afdc913869

> Indicates support for transferring files that are larger than 4,294,967,295 bytes in size. If this flag is not set, then only files of size less than or equal to 4,294,967,295 bytes can be exchanged using the File Contents Request PDU and File Contents Response PDU.

Large file (4GiB) support will be implemented as an extension and indicated by `CB_HUGE_FILE_SUPPORT_ENABLED`.  

https://github.com/neutrinolabs/xrdp/blob/b367f57036f9c9ac2501eb83f7cafe7b0bbaeaef/sesman/chansrv/clipboard.c#L238-L240

However, `g_file_get_size()` returns file size as signed int. So the first goal is to extend 2GiB file size limit to 4GiB.

**Reply by matt335672** [reply]
Sounds reasonable.

We'll still need to pass an error back from `g_get_file_size()` somehow though, if we don't use `off_t` or something like it.
